# Mahindra Club Resort Spending Prediction


In [149]:
import numpy as np 
import pandas as pd

import matplotlib.pyplot as plt
import seaborn as sns

from tqdm import tqdm

import warnings
warnings.simplefilter("ignore")


### Loading data and High level info of data


In [150]:
data = pd.read_csv('./train_clubmahindra.csv')
data

reservation_id booking_date  \
0       07659f3758d8aee27f5a7e2887adeacb67021cb95ada1b...     05/04/18   
1       03930f033646d073462b35d411616323597715ac4fc398...     23/01/15   
2       d145a32920e6587ad95bfe299d80c0affa268220535aaf...     28/01/15   
3       cfd77f44811ed62f25a220b53324cdbafc662a4c9e5f04...     02/05/15   
4       937cff9e4dcfc2459620153dfc8b9962ac22bea67dfb29...     02/09/15   
...                                                   ...          ...   
341419  a804d59b74b1d532a90740ded80dd05873bbef125bc349...     11/02/19   
341420  7062c92a631731e7189cf3dddb20aceeae91a883e14229...     27/02/19   
341421  18a5668c302c321dc2d42a48923ceff97af633c66d72ee...     27/02/19   
341422  1cbe1611a122023d7e994b4cf51fcf28ce44acdbcabbfb...     27/02/19   
341423  8712dd3ce2e36f2fbcd16b24aa3b89bd1c5203a39e578b...     28/02/19   

       checkin_date checkout_date  channel_code  main_product_code  \
0          05/04/18      06/04/18             3                  1   
1          11/04/15      16/04/15             1                  1   
2          01/02/15      05/02/15             1                  1   
3          11/06/15      16/06/15             1                  1   
4          14/12/15      19/12/15             1                  1   
...             ...           ...           ...                ...   
341419     27/02/19      01/03/19             2                  2   
341420     28/02/19      01/03/19             2                  4   
341421     27/02/19      28/02/19             3                  3   
341422     28/02/19      01/03/19             1                  3   
341423     01/03/19      02/03/19             3                  3   

        numberofadults  numberofchildren  persontravellingid  \
0                    2                 0                  46   
1                    2                 0                  46   
2                    2                 0                  47   
3                    2                 2                  46   
4                    2                 0                  46   
...                ...               ...                 ...   
341419               4                 0                  47   
341420               2                 0                  45   
341421               3                 0                  45   
341422               5                 0                  47   
341423               2                 0                  45   

        resort_region_code  ...  state_code_residence  state_code_resort  \
0                        3  ...                   7.0                  3   
1                        3  ...                   7.0                  5   
2                        1  ...                   7.0                  1   
3                        2  ...                   7.0                  2   
4                        2  ...                   7.0                  2   
...                    ...  ...                   ...                ...   
341419                   1  ...                   8.0                  1   
341420                   1  ...                  10.0                  1   
341421                   1  ...                  10.0                  4   
341422                   2  ...                   8.0                  2   
341423                   2  ...                   8.0                  7   

        total_pax  member_age_buckets  booking_type_code  \
0               3                   F                  1   
1               2                   F                  1   
2               2                   F                  1   
3               2                   F                  1   
4               2                   F                  1   
...           ...                 ...                ...   
341419          2                   I                  1   
341420          2                   D                  1   
341421          3                   H                  1   
341422          3              

In [151]:
data.describe()

channel_code  main_product_code  numberofadults  numberofchildren  \
count  341424.000000      341424.000000   341424.000000     341424.000000   
mean        1.902810           2.208738        3.275361          0.362696   
std         0.915311           0.823779        1.756062          0.758019   
min         1.000000           1.000000        0.000000          0.000000   
25%         1.000000           2.000000        2.000000          0.000000   
50%         2.000000           2.000000        3.000000          0.000000   
75%         3.000000           3.000000        4.000000          0.000000   
max         3.000000           7.000000       32.000000         13.000000   

       persontravellingid  resort_region_code  resort_type_code  \
count       341424.000000       341424.000000     341424.000000   
mean           210.288905            1.772960          2.193832   
std            865.735577            0.750806          1.551451   
min             45.000000            1.000000          0.000000   
25%             45.000000            1.000000          1.000000   
50%             45.000000            2.000000          2.000000   
75%             45.000000            2.000000          3.000000   
max           4995.000000            3.000000          7.000000   

       room_type_booked_code     roomnights  season_holidayed_code  \
count          341424.000000  341424.000000          341310.000000   
mean                2.890321       3.738346               2.673862   
std                 0.804260       2.479054               0.885650   
min                 1.000000     -45.000000               1.000000   
25%                 2.000000       2.000000               2.000000   
50%                 3.000000       3.000000               3.000000   
75%                 3.000000       4.000000               3.000000   
max                 6.000000      60.000000               4.000000   

       state_code_residence  state_code_resort      total_pax  \
count         336660.000000      341424.000000  341424.000000   
mean               7.260708           4.763303       3.193390   
std                4.701283           3.103909       1.167824   
min                1.000000           1.000000       0.000000   
25%                4.000000           2.000000       2.000000   
50%                8.000000           5.000000       3.000000   
75%                8.000000           7.000000       4.000000   
max               38.000000          13.000000      24.000000   

       booking_type_code  amount_spent_per_room_night_scaled  
count      341424.000000                       341424.000000  
mean            1.124209                            7.720801  
std             0.329821                            1.090109  
min             1.000000                            1.600397  
25%             1.000000                            7.187260  
50%             1.000000                            7.860814  
75%             1.000000                            8.419144  
max             2.000000                           10.816652

In [152]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 341424 entries, 0 to 341423
Data columns (total 24 columns):
 #   Column                              Non-Null Count   Dtype  
---  ------                              --------------   -----  
 0   reservation_id                      341424 non-null  object 
 1   booking_date                        341424 non-null  object 
 2   checkin_date                        341424 non-null  object 
 3   checkout_date                       341424 non-null  object 
 4   channel_code                        341424 non-null  int64  
 5   main_product_code                   341424 non-null  int64  
 6   numberofadults                      341424 non-null  int64  
 7   numberofchildren                    341424 non-null  int64  
 8   persontravellingid                  341424 non-null  int64  
 9   resort_region_code                  341424 non-null  int64  
 10  resort_type_code                    341424 non-null  int64  
 11  room_type_booked_code     

## Pre-processing data

## Filling missing values

There are few missing values in following columns 
- season_holidayed_code               
- state_code_residence                

Even though both of them are floating data actually from the observation the are whole numbers with a finite set of unique values which seems to be representing some categorical features. So, replacing this missing values with mode.

In [153]:
data['season_holidayed_code'].fillna(data['season_holidayed_code'].mode()[0], inplace= True)
data['state_code_residence'].fillna(data['state_code_residence'].mode()[0], inplace= True)

data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 341424 entries, 0 to 341423
Data columns (total 24 columns):
 #   Column                              Non-Null Count   Dtype  
---  ------                              --------------   -----  
 0   reservation_id                      341424 non-null  object 
 1   booking_date                        341424 non-null  object 
 2   checkin_date                        341424 non-null  object 
 3   checkout_date                       341424 non-null  object 
 4   channel_code                        341424 non-null  int64  
 5   main_product_code                   341424 non-null  int64  
 6   numberofadults                      341424 non-null  int64  
 7   numberofchildren                    341424 non-null  int64  
 8   persontravellingid                  341424 non-null  int64  
 9   resort_region_code                  341424 non-null  int64  
 10  resort_type_code                    341424 non-null  int64  
 11  room_type_booked_code     

Now there are no null values in the data set.

### Removing id information

Dropping columns related to id such as reservation_id, persontravellingid and memeberid, because there is a high possibility getting new id data in the real time for new dataset

In [154]:
data.drop(columns= ['reservation_id', 'persontravellingid', 'memberid', 'resort_id'], inplace= True)

### Pre-processing on date columns

In [155]:
date_columns = data.filter(regex = '.*date$').copy()

for column in date_columns.columns:
    date_columns[column] = pd.to_datetime(date_columns[column])

In [156]:
date_columns.loc[date_columns['checkin_date'] > date_columns['checkout_date']]

booking_date checkin_date checkout_date
1        2015-01-23   2015-11-04    2015-04-16
3        2015-02-05   2015-11-06    2015-06-16
5        2015-01-12   2016-12-01    2016-01-13
10       2017-03-02   2017-02-28    2017-01-03
34       2018-03-31   2012-12-03    2012-03-18
...             ...          ...           ...
341406   2019-02-20   2019-02-27    2019-02-03
341417   2019-02-22   2019-02-27    2019-02-03
341419   2019-11-02   2019-02-27    2019-01-03
341420   2019-02-27   2019-02-28    2019-01-03
341422   2019-02-27   2019-02-28    2019-01-03

[46126 rows x 3 columns]

Above show is data where checkin date is after checkout date and from the observation the months and dates are swapped in chekin or checkout dates which is this dataset and the same is absovrved when we see this abnormality is seen in between booking date with the other two dates. Considering this huge error in the collection of dates **its better to drop date related columns from our dataset.**

In [157]:
data.drop(columns=['booking_date', 'checkin_date', 'checkout_date'], inplace= True)

### Pre-processing non-numeric data

In [158]:
categorical_data = data.select_dtypes(exclude = 'number')

data.drop(columns= categorical_data.columns, inplace= True)
categorical_data.nunique()

member_age_buckets          10
cluster_code                 6
reservationstatusid_code     4
dtype: int64

There are small number of columns so using one hot encoding for transforming this categorical features into numeric ones

In [159]:
encoded_categorical_data = pd.get_dummies(categorical_data)

data = pd.concat([data, encoded_categorical_data], axis= 1)

In [160]:
data.to_csv('./pre-processed-data.csv')

now we have 341424 rows × 34 columns of pure numeric data and we can proceed to the data analysis phase.

## Data analysis

## Machine Learning

In [161]:
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import Normalizer

In [162]:
normalizer = Normalizer()


In [163]:
y = data['amount_spent_per_room_night_scaled']
X = normalizer.fit_transform(data.drop(columns= ['amount_spent_per_room_night_scaled']))

In [164]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size= 0.2, random_state= 41)

In [165]:
model = LinearRegression()
model.fit(X_train, y_train)
predicted = model.predict(X_test)

In [166]:
mean_squared_error(y_test, predicted)

1.116636997528317

### Ridge

In [167]:
model = Ridge(alpha=0.001)

model.fit(X_train, y_train)

predicted = model.predict(X_test)


In [168]:
mean_squared_error(y_test, predicted)

1.1166367691202035

### Lasso

In [169]:
model = Lasso(alpha=0.001)

model.fit(X_train, y_train)

predicted = model.predict(X_test)


In [170]:
mean_squared_error(y_test, predicted)

1.1253403577695504